lib/spack/docs/tutorial_advanced_packaging.rst

```diff
@@ -0,0 +1,560 @@
+  ==> Successfully installed netcdf
+    Fetch: 0.01s.  Build: 24.61s.  Total: 24.62s.
+  [+] /usr/local/opt/spack/linux-ubuntu16.04-x86_64/gcc-5.4.0/netcdf-4.4.1.1-gk2xxhbqijnrdwicawawcll4t3c7dvoj
+
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Single package providing multiple virtual specs
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+At the close of this tutorial's subsection, it may be useful to see where the
+build-interface protocol shines the most i.e. when it comes to manage packages
+that provide more than one virtual spec. An example of a package of this kind is
+``intel-parallel-studio``, and due to its complexity we'll limit our discussion
+here to just a few considerations (without any hands-on). You can open
+the related ``package.py`` in the usual way:
+
+.. code-block:: console
+
+  root@advanced-packaging-tutorial:/# spack edit intel-parallel-studio
+
+As you can see this package provides a lot of virtual specs, and thus it has
+more than one function that enters into the build-interface protocol. These
+functions will be invoked for *exactly the same spec* according to the key used
+by its dependents in the subscript query.
+
+So, for instance, the ``blas_libs`` property will be returned when
+``intel-parallel-studio`` is the BLAS provider in the current DAG and
+is retrieved by a dependent with:
+
+.. code-block:: python
+
+  blas = self.spec['blas']
+
+Within the property we inspect various aspects of the current spec:
+
+.. code-block:: python
+
+  @property
+  def blas_libs(self):
+     spec = self.spec
+     prefix = self.prefix
+     shared = '+shared' in spec
+
+     if '+ilp64' in spec:
+         mkl_integer = ['libmkl_intel_ilp64']
+     else:
+         mkl_integer = ['libmkl_intel_lp64']
+     ...
+
+and construct the list of library we need to return accordingly.
+
+What we achieved is that the complexity of dealing with ``intel-parallel-studio``
+is now gathered in the package itself, instead of being spread
+all over its possible dependents. A package using MPI, or LAPACK, thus don't care
+anymore what is the actual implementation that is going to be used, and has
+*a uniform interface* to ask for libraries or headers and manipulate them.
+The packages that provide this virtual spec, on the other hand, have a clear
+way to differentiate their answer to the query.
+
+---------------------------
+Package's build environment
+---------------------------
+
+Besides Spec's build interface, Spack provides means to set environment
+variables, either for yourself or for your dependent packages, and to
+attach attributes to your dependents. We'll see them next with the help
+of a few real use cases.
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Set variables at build-time for yourself
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A few packages require their users to set some environment
+variables in order to be configured and built correctly. Spack
+permits you to accomplish this using the
```

Comment:
I'd like to see the ability to independently tweak the build and runtime environments called out earlier, e.g.

"Spack provides a way to manipulate a package's build time and run time environments using the `setup_environment` function.